### ISSUE TITLE
fatal: ambiguous argument '2021.11.29^': unknown revision or path not in the working tree.

### ISSUE THREAD

**Issue by ReLater** [DigiLive/gitChangelog#14]
**Describe the bug**
- Version 1.0.1
- Win 10 Pro, WSL(Debian)-Linux, GitHub repository.
- After adding the changes from https://github.com/DigiLive/gitChangelog/pull/13
- Run a custom PHP script with changed gitPath (but I think not relevant here):

```
<?php
use DigiLive\GitChangelog\Renderers\MarkDown;

require __DIR__ . '/vendor/autoload.php';

// Instantiate the library's renderer.
$changelog = new MarkDown('../');

// Build and save the changelog with all defaults.
$changelog->build();
$changelog->save('CHANGELOG.md');
```
- Error during fetching tags:
```
fatal: ambiguous argument '2021.11.29^': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
fatal: ambiguous argument '2021.11.29^': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
PHP Fatal error:  Uncaught RuntimeException: An error occurred while fetching the commit data from the repository. in vendor/digilive/git-changelog/src/GitChangelog.php:293
Stack trace:
#0 vendor/digilive/git-changelog/src/Renderers/MarkDown.php(86): DigiLive\GitChangelog\GitChangelog->fetchCommitData()
#1 git-changelog.php(16): DigiLive\GitChangelog\Renderers\MarkDown->build()
#2 {main}
  thrown in vendor/digilive/git-changelog/src/GitChangelog.php on line 293
```
**Investigations**
- I have a tag commit named `2021.11.29` in directory `.git/refs/tags`.
- GitChangelog.php adds a `^` to the $tag on line 275 (`$tagRange   = $rangeStart !== false ? "$rangeStart..$tag" : "$tag^";`).

**Solution (but I'm a complete git idiot; just trial & error)**
- When I remove `^` everything works correctly and I get a clean CHANGELOG.md (`$tagRange   = $rangeStart !== false ? "$rangeStart..$tag" : "$tag";`)

```
...
## 2021.11.29 (2021-11-29)`
...
```

**Comment by JBlond** [comment]
@DigiLive This was done in https://github.com/DigiLive/gitChangelog/commit/fc79a583e2cd32d78879ecc720e5cc21fba96f39

**Comment by ReLater** [comment]
Just an addition:

> I have a tag commit named 2021.11.29 in directory .git/refs/tags.

It's the very first tag in that repository.

**Comment by DigiLive** [comment]
Given error most likely occurs when the tag doesn't exist in the repo.

I would like to know the output of the following commands:
`git version`
`git --git-dir ../.git tag`
`git --git-dir ../.git log 2021.11.29 --pretty=format:%s`
`git --git-dir ../.git log 2021.11.29^ --pretty=format:%s`

Run these command at a cli from the same directory as your script.

**Comment by ReLater** [comment]
- `$ git version`
=>
```
git version 2.20.1
```
- `$ git --git-dir ../.git tag`
=>
```
2021.11.29
2021.12.08
2022.01.03
2022.01.12
2022.01.14
2022.01.27
2022.01.27.2
2022.01.30
2022.02.07
2022.02.08
2022.02.12
2022.02.15
2022.02.25
2022.02.27
2022.03.07
2022.03.13
```
- `$ git --git-dir ../.git log 2021.11.29 --pretty=format:%s`
=>
```
Initial commit
```
- `$ git --git-dir ../.git log 2021.11.29^ --pretty=format:%s`
=>
```
fatal: ambiguous argument '2021.11.29^': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
```

**Comment by ReLater** [comment]
Same result with `git version 2.35.1`

**Comment by DigiLive** [comment]
I'm afraid I can't reproduce the error.
Would you mind sharing your `'../.git'` folder with me?

You can contact me at github [at] digilive.nl

**Comment by ReLater** [comment]
I will invite you to the private repository.

Maybe relevant: The local repositories are on a WIN harddisc `Z:/`. I access them from WSL-Debian via `/mnt/z/`.

**Comment by ReLater** [comment]
But please don't waste too much time with this issue. I have tested your library in a similar repository in identical local environment. There it works.

Only difference as far as I see: The first tag has not a single `Initial commit` but several "real" commits.

**Comment by ReLater** [comment]
The repository is public now for a while. You can clone/fork it.

- All relevant PHP is in folder `/buildBefore/`.
- `/buildBefore/vendor/digilive/git-changelog/src/GitChangelog.php` is a copy of your current `main` branch.
- My script is `/buildBefore/git-changelog.php`
- The whole repository is chaotic. A playground ;-)

https://github.com/GHSVS-de/hugo_docsghsvs

**Comment by DigiLive** [comment]
> But please don't waste too much time with this issue.

As a software developer, I have to ;)
I want my packages to be bug-free so they can be used with pleasure.

---

I've been able to reproduce the issue now.
There are differences between running git in Windows CMD shell and PowerShell/Bash.
CMD: no differences between `tag^` and `tag`
PS/Bash: `tag^` _excludes_ tagged commit and `tag` _includes_ tagged commit.

I assume you're executing php in a shell similar to PS/Bash, which will raise the error for your repo.
Unfortunately in this case, my php binaries are for Windows as I don't have Linux at my disposal.

Versions used:
```
Win10:   21H1 build 19043.1526
git:     2.35.1.windows.2
CMD:     10.0.19043.1526
PS:      5.1.19041.1320
Bash: GNU bash, version 4.4.23(2)-release (x86_64-pc-msys)
```
---
Because `2021.11.29^` excludes the tagged commit and your repo doesn't have any earlier commits than this tagged one, the error is raised.
In this case, the important part of the error is `unknown revision`. I.e. There is no commit which matches the criteria.

Thank you very much for posting this issue!
Without your help, I would never have detected this bug.

I really have to re-think why I've added the caret, but I guess it can be safely removed.

**Note to self: Set Bash as default shell for this package.**

**Comment by ReLater** [comment]
I use WSL1 with Debian, which is a fully functinal console Linux but I have absolutely no idea which Shell/Bash/whatever it uses finally when I click the Debian icon. I could not find any hint but "conhost.exe" (Windows console host).

**Comment by DigiLive** [comment]
It's probably 
Bash: An interactive shell is used by a user in a virtual console or TerminalEmulator.  Installed at `/etc/shells`
or 
Dash: A non-interactive shell is used to execute system scripts or scripts that use #!/bin/sh.

**Comment by JBlond** [comment]
WSL by default uses bash when logged in as a user

**Comment by DigiLive** [comment]
Fixed in [c96a476](https://github.com/DigiLive/gitChangelog/commit/c96a4768d6d33ce3ffd7b9d9dede274e2a105762) of the main branch.
